### PR TITLE
Call apiv4 from Contribution create rather than fugly addActivity function

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------

Use apiv4 to create activity from Contribution create (alternative to https://github.com/civicrm/civicrm-core/pull/17274 which has blocking test failures).

Before
----------------------------------------
Hard-to-read & largely pointless function called which can hide fails

After
----------------------------------------
APIV4 called

Technical Details
----------------------------------------
Generally APIv4 is where we want to go - additional reasons in https://github.com/civicrm/civicrm-core/pull/17274

 I looked at  - but felt that #17274 
the shared function was adding nothing and simply using the api to create the activity made more sense. The
shared function does a lot of silly wrangling for very little shared functionality and is
hard to read. In this call only 2 params are passed in - so most of the wranglingg
doesn't apply anyway. I ensured the 2 JIRA issues referenced in the removed code have test cover (one already
had a test written by Monish & I added in the campaign check - see #8384 & #1775

Also this is only affecting one of the ways the function is called so it should also be less risky

Comments
----------------------------------------

@colemanw does this look right to you

### CODE REVIEW

**Comment by mattwire** [comment]
@eileenmcnaughton Test failures look related:
```
ERROR USERINFO: INSERT INTO `civicrm_activity` (`source_record_id` , `activity_type_id` , `subject` , `activity_date_time` , `status_id` , `priority_id` , `is_test` , `campaign_id` ) VALUES ( 1 ,  6 , '$ 77.00 - AnnualFixed Membership Signup: Credit card or direct debit (by Mr. Logged In User 761816242 II)' ,  20200718070724 ,  2 ,  2 ,  1 ,  0 )  [nativecode=1452 ** Cannot add or update a child row: a foreign key constraint fails (`core1788157jn8test_7flvb`.`civicrm_activity`, CONSTRAINT `FK_civicrm_activity_campaign_id` FOREIGN KEY (`campaign_id`) REFERENCES `civicrm_campaign` (`id`) ON DELETE SET NULL)]
```

File: CRM/Contribute/BAO/Contribution.php [hunk 1]
```diff
@@ -512,26 +515,33 @@ public static function create(&$params, $ids = []) {
...
+
+      $campaignParams = isset($params['campaign_id']) ? ['campaign_id' => ($params['campaign_id'] ?? NULL)] : [];
+      Activity::save(FALSE)->addRecord(array_merge([
+        'activity_type_id:name' => 'Contribution',
+        'source_record_id' => $contribution->id,
+        'source_contact_id' => CRM_Core_Session::getLoggedInContactID() ?: $contribution->contact_id,
+        'target_contact_id' => CRM_Core_Session::getLoggedInContactID() ? [$contribution->contact_id] : [],
+        'activity_date_time' => $contribution->receive_date,
+        'is_test' => (bool) $contribution->is_test,
+        'status_id.name' => $isCompleted ? 'Completed' : 'Scheduled',
```
Comment (colemanw) [inline CRM/Contribute/BAO/Contribution.php:539]: I think this is a typo.

```suggestion
        'status_id:name' => $isCompleted ? 'Completed' : 'Scheduled',
```

File: tests/phpunit/CRM/Contribute/BAO/ContributionTest.php [hunk 2]
```diff
@@ -886,11 +887,10 @@ public function testCheckLineItemsWithFloatingPointValues() {
   }
 
   /**
-   * Test activity amount updation.
+   * Test activity amount updates activity subject.
```
Comment (colemanw) [inline tests/phpunit/CRM/Contribute/BAO/ContributionTest.php:890]: LOL, you mean updation isn't a word?

File: tests/phpunit/CRM/Contribute/BAO/ContributionTest.php [hunk 3]
```diff
@@ -913,31 +912,30 @@ public function testActivityCreate() {
...
+    $params['campaign_id'] = $this->campaignCreate();
 
     $contribution = $this->callAPISuccess('Contribution', 'create', $params)['values'][0];
 
     $this->assertEquals($params['total_amount'], $contribution['total_amount'], 'Check for total amount in contribution.');
     $this->assertEquals($contactId, $contribution['contact_id'], 'Check for contact id  creation.');
 
     // Retrieve activity again.
-    $activity = CRM_Activity_BAO_Activity::retrieve($activityParams, $defaults);
+    $activity = Activity::get()->setWhere($activityWhere)->setSelect(['source_record_id', 'subject', 'campaign_id'])->execute()->first();
```
Comment (colemanw) [inline tests/phpunit/CRM/Contribute/BAO/ContributionTest.php:934]: $checkPermissions = TRUE is intentional here?

**Review by colemanw — CHANGES_REQUESTED** [review]
Check possible typo

File: tests/phpunit/CRM/Contribute/BAO/ContributionTest.php (hunk 3 above)
Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Contribute/BAO/ContributionTest.php:934]: @colemanw I think in tests we generally don't need to

File: CRM/Contribute/BAO/Contribution.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Contribute/BAO/Contribution.php]: @colemanw thanks! fixed

**Review by colemanw — APPROVED** [review]

**Comment by colemanw** [comment]
Merging based on code review & tests.

**Pull request merged**